### PR TITLE
[ASA] Add Teorema Mestre

### PR DESCRIPTION
**Pull request by Calhau18**

### CODE REVIEW

**Comment by randomicecube** [comment]
Still missing examples - waiting for the teacher's notes.

**Comment by randomicecube** [comment]
This commit also needs the slides to be added (leaving this here so I don't forget to add them)

File: content/asa/0002-teorema-mestre.md [hunk 1]
```diff
@@ -0,0 +1,522 @@
...
+g \in O(f) \text{ sse } \exists_{c}\exists_{n_0}, \forall_{n \geq n_0}, \quad g(n) \leq c\cdot f(n)
+$$
+
+É o **limite assimptótico superior**, com notação $O$ - afere a complexidade no pior caso.
+
+:::details[**Limite assimptótico superior**]
+
+![Limite Assimptótico Superior](./assets/0002-maj-asympt.png)
+
+:::
```
Comment (diogotcorreia) [inline content/asa/0002-teorema-mestre.md:32]: Should be indented to be part of the bullet list

File: content/asa/0002-teorema-mestre.md [hunk 2]
```diff
@@ -0,0 +1,522 @@
...
+
+- Minorante assimptótico:
+
+$$
+g \in \Omega(f) \text{ sse } \exists_{c}\exists_{n_0}, \forall_{n \geq n_0}, \quad g(n) \geq c\cdot f(n)
+$$
+
+É o **limite assimptótico inferior**, com notação $\Omega$ - afere a complexidade no melhor caso.
```
Comment (diogotcorreia) [inline content/asa/0002-teorema-mestre.md:40]: Should be indented to be part of the bullet list

File: content/asa/0002-teorema-mestre.md [hunk 3]
```diff
@@ -0,0 +1,522 @@
...
+g \in \Theta(f) \text{ sse } \exists_{c_1, c_2}\exists_{x_0}, \forall_{n \geq n_0}, \quad c_1\cdot f(n) \leq g(n) \leq c_2\cdot f(n)
+$$
+
+É o **limite assimptótico apertado**, com notação $\Theta$ - quando o melhor e o pior caso têm a mesma complexidade.
+
+:::details[**Limite Assimptótico Apertado**]
+
+![Limite Assimptótico Apertado](./assets/0002-tight-asympt.png)
+
+:::
```
Comment (diogotcorreia) [inline content/asa/0002-teorema-mestre.md:54]: Should be indented to be part of the bullet list

File: content/asa/0002-teorema-mestre.md [hunk 4]
```diff
@@ -0,0 +1,522 @@
...
+f \in O(g) \wedge g \in O(h) \implies f \in O(h).
+$$
+
+:::
+
+[**As provas de ambos os lemas encontram-se nas notas do professor, no fim da página.**](color:green)
+
+---
+
+Muitas vezes, uma abordagem que permite diminuir significativamente o tempo assimptótico em que é possível resolver um problema é usar uma abordagem de [**Dividir e conquistar**](color:orange).
```
Comment (diogotcorreia) [inline content/asa/0002-teorema-mestre.md:78]:
```suggestion
Muitas vezes, uma abordagem que permite diminuir significativamente o tempo assimptótico em que é possível resolver um problema é usar uma abordagem de [**Dividir e Conquistar**](color:orange).
```

File: content/asa/0002-teorema-mestre.md [hunk 5]
```diff
@@ -0,0 +1,522 @@
...
+
+Metodologia:
+
+- [**D**](color:orange)ividir o problema a resolver num conjunto de subproblemas.
+
+- [**R**](color:yellow)esolver (recursivamente) cada um dos subproblemas.
+
+- [**C**](color:red)ombinar as soluções dos subproblemas para obter a solução do problema original.
+
+![Dividir e Conquistar](./assets/0002-divide-conquer.png)
```
Comment (diogotcorreia) [inline content/asa/0002-teorema-mestre.md:90]: I'm pretty sure `C` and  `R` are accidentally swapped on this image

File: content/asa/0002-teorema-mestre.md [hunk 6]
```diff
@@ -0,0 +1,522 @@
...
+
+Fica então fácil de generalizar que no nível $k$ teremos $a^k$ problemas de tamanho $\frac{n}{b^k}$ com custo pontual $O(\frac{n}{b^k}^d)$.
+
+Calcular a complexidade da nossa solução corresponde então a somar a complexidade de cada nível, até um nível da árvore em que o custo pontual é constante (neste caso assumimos que isso acontece apenas quando $n=1$).
+Para isso precisamos de saber quantas divisões temos de fazer até chegar a esse nível. A resposta é o valor $k$ tal que $\frac{n}{b^k} = 1 \Leftrightarrow k = \log_b n$.
+
+Ficamos então com o somatório:
+
+$$
+\sum_{k=0}^{\log_b n} O \left( a^k (\frac{n}{b^k})^d \right) =
```
Comment (diogotcorreia) [inline content/asa/0002-teorema-mestre.md:149]:
```suggestion
\sum_{k=0}^{\log_b n} O \left( a^k \left(\frac{n}{b^k}\right)^d \right) =
```

File: content/asa/0002-teorema-mestre.md [hunk 7]
```diff
@@ -0,0 +1,522 @@
...
+$$
+2k \geq n \leftrightarrow k \geq \frac{n}{2}
+$$
+
+Podemos, então, dizer que o ciclo corre $\frac{n}{2}$ vezes, e que a sua complexidade temporal é $O(n)$.
+
+Como observado, $j = k$ no fim do ciclo, pelo que a chamada recursiva da função (dentro do bloco `if`) é realizada com [$j = \frac{n}{2}$](color:orange) [**duas vezes**](color:yellow) (a função recursiva é chamada duas vezes), pelo que podemos admitir que essa secção leva $2\cdot T(\frac{n}{2})$. A função $T(n)$ total da função corresponde, portanto, a
+
+$$
+T(n) = \textcolor{yellow}{2}\cdot T\left(\textcolor{orange}{\frac{n}{2}}\right) + O(n).
```
Comment (diogotcorreia) [inline content/asa/0002-teorema-mestre.md:237]: Colors on KaTeX have a problem that they're not dynamic, that is, they don't adapt to the light/dark theme of the website. I have a solution in mind for this, I'll fix it myself.

File: content/asa/0002-teorema-mestre.md [hunk 8]
```diff
@@ -0,0 +1,522 @@
...
+$$
+\forall_{1 < i \leq n}, \quad A[parent(i)] \geq A[i],
+$$
+
+com
+
+$$
+parent(i) = \left\lfloor{\frac{i}{2}}\right\rfloor\\
+left(i) = 2i\\
+right(i) = 2i + 1
```
Comment (diogotcorreia) [inline content/asa/0002-teorema-mestre.md:502]:
```suggestion
\forall_{1 < i \leq n}, \quad A[\operatorname{parent}(i)] \geq A[i],
$$

com

$$
\operatorname{parent}(i) = \left\lfloor{\frac{i}{2}}\right\rfloor\\
\operatorname{left}(i) = 2i\\
\operatorname{right}(i) = 2i + 1
```

**Review by diogotcorreia — CHANGES_REQUESTED** [review]
I'll take care of this

**Comment by Pesteves2002** [comment]
> I'll take care of this

"Fine, I'll do it myself"

**Review by diogotcorreia — APPROVED** [review]

**Pull request merged**